### PR TITLE
iproto: add new keys to Binary protocol page

### LINKED ISSUE #3166: iproto: new key - IPROTO_VCLOCK_SYNC
**Product:** Tarantool
**Since:** 2.11
**Root document:** https://www.tarantool.io/en/doc/latest/dev_guide/internals/box_protocol/
https://www.tarantool.io/en/doc/latest/dev_guide/internals/box_protocol/#heartbeats
**SME:** @ sergepetrenko
**Estimation:** 2sp

# Details

Binary protocol receives a new key:
IPROTO_VCLOCK_SYNC = 0x5a
This key holds a MP_UINT value and is used only by replication
heartbeats. A master sends the monotonically growing vclock sync together
with some of its heartbeats, and replica replies with the greatest vclock
sync it has seen yet in its ACKs.
Requested by @sergepetrenko in https://github.com/tarantool/tarantool/commit/e00f9dad385d02417ea0ddfedb7c668c20b420db.

### LINKED ISSUE #3193: iproto: new field in applier -> relay ACKs
**Product:** Tarantool
**Since:** 2.11
**Audience/target:** dev
**Root document:** https://www.tarantool.io/en/doc/latest/dev_guide/internals/box_protocol/
**SME:** @ Gerold103

# Details

Do this after https://github.com/tarantool/doc/pull/3151 is merged.
Update the Replication and Keys sections.

---
The applier->relay channel (from replica back to master) is used
only for sending ACKs. Replication data goes the other way
(relay->applier).

These ACKs had 2 fields: `IPROTO_VCLOCK (0x26)` and
`IPROTO_VCLOCK_SYNC (0x5a)`.

Now they have a new field: `IPROTO_TERM (0x53)`. It is a unsigned
number containing `box.info.election.term` of the sender node
(applier, replica).
Requested by @Gerold103 in https://github.com/tarantool/tarantool/commit/544955109140e5dd5b956178fe47fd5ed3da2952.

### LINKED ISSUE #3294: iproto: new fields in instance's ballot.
**Product:** Tarantool
**Since:** 2.11
**Root document:** https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/replication/#box-protocol-ballots
https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general-replication
**SME:** @ sergepetrenko

# Details

Instance's ballot (a response to IPROTO_VOTE sent on replica connect)
receives two new fields:
1) The uuid of the node this instance considers the bootstrap leader.
   Key: IPROTO_BALLOT_BOOTSTRAP_LEADER_UUID = 0x08
   Value: uuid, encoded as 36-byte string (like
   "bfd2b31c-b740-43e5-bf3c-28538a74c9a6").
2) An array of registered replica set members uuids.
   Key: IPROTO_BALLOT_REGISTERED_REPLICA_UUIDS = 0x09
   Value: a MP_ARRAY of uuids, each uuid encoded as a 36-byte string
   (like in an example above).
Requested by @sergepetrenko in https://github.com/tarantool/tarantool/commit/fd61dc64b91524c47fa9b8652fb1b05dab968297.

### LINKED ISSUE #3299: iproto: rename `IPROTO_CLUSTER_UUID` to `IPROTO_REPLICASET_UUID`
**Product:** Tarantool
**Since:** 2.11
**Audience/target:** dev
**Root document:** https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general-replication
**SME:** @ Gerold103 

# Details
This is a name for one of the IProto keys. The key value doesn't
change and the protocol is still backward compatible. But better
rename it to `IPROTO_REPLICASET_UUID`, because in future
`IPROTO_CLUSTER_UUID` will most likely mean a different thing.
Requested by @Gerold103 in https://github.com/tarantool/tarantool/commit/a24714c55305280838576fa6dec9a9781763e893.

### PR DESCRIPTION
**Pull request by xuniq**
Resolves https://github.com/tarantool/doc/issues/3166
Resolves https://github.com/tarantool/doc/issues/3294
Resolves https://github.com/tarantool/doc/issues/3193
Resolves https://github.com/tarantool/doc/issues/3299

1) Updated keys in [General Replication](https://docs.d.tarantool.io/en/doc/gh-3166-iproto-keys/dev_guide/internals/iproto/keys/#general-replication) section:
* Added IPROTO_VCLOCK_SYNC key
* Renamed IPROTO_CLUSTER_UUID key to IPROTO_REPLICASET_UUID
* Added IPROTO_BALLOT_BOOTSTRAP_LEADER_UUID and IPROTO_BALLOT_REGISTERED_REPLICA_UUIDS keys

2) Updated [vclock keys](https://docs.d.tarantool.io/en/doc/gh-3166-iproto-keys/dev_guide/internals/iproto/keys/#vclock-keys) description

3) Updated [description and svg for heartbeat messages](https://docs.d.tarantool.io/en/doc/gh-3166-iproto-keys/dev_guide/internals/iproto/replication/#heartbeats) (added IPROTO_VCLOCK_SYNC and IPROTO_TERM)

4) Updated IPROTO_BALLOT description

* https://docs.d.tarantool.io/en/doc/gh-3166-iproto-keys/dev_guide/internals/iproto/keys/#iproto-ballot-keys
* https://docs.d.tarantool.io/en/doc/gh-3166-iproto-keys/dev_guide/internals/iproto/replication/#iproto-ballot

5) Updated IPROTO_TERM key description:
* https://docs.d.tarantool.io/en/doc/gh-3166-iproto-keys/dev_guide/internals/iproto/keys/#synchronous-replication
* https://docs.d.tarantool.io/en/doc/gh-3166-iproto-keys/dev_guide/internals/iproto/keys/#iproto-term

6) [Updated byte code examples for the heartbeat messages](https://docs.d.tarantool.io/en/doc/gh-3166-iproto-keys/how-to/other/iproto/#heartbeat) (added IPROTO_VCLOCK_SYNC and IPROTO_TERM)

### CODE REVIEW

File: doc/dev_guide/internals/iproto/images/repl_ballot.puml [hunk 1]
```diff
@@ -8,13 +8,15 @@ skinparam {
...
+    "[[https://tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general-replication IPROTO_BALLOT_VCLOCK]]": "MP_MAP",
+    "[[https://tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general-replication IPROTO_BALLOT_GC_VCLOCK]]": "MP_MAP",
+    "[[https://tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general-replication IPROTO_BALLOT_IS_RO]]": "MP_BOOL",
+    "[[https://tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general-replication IPROTO_BALLOT_IS_ANON]]": "MP_BOOL",
+    "[[https://tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general-replication IPROTO_BALLOT_IS_BOOTED]]": "MP_BOOL",
+    "[[https://tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general-replication IPROTO_BALLOT_CAN_LEAD]]": "MP_BOOL",
+    "[[https://tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general-replication IPROTO_BALLOT_BOOTSTRAP_LEADER_UUID]]": "MP_STR",
+    "[[https://tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general-replication IPROTO_BALLOT_REGISTERED_REPLICA_UUIDS]]": "MP_ARRAY"
```
Comment (sergepetrenko) [inline doc/dev_guide/internals/iproto/images/repl_ballot.puml:19]: Should we describe the types of array elements somehow?

Like "MP_ARRAY of MP_STR elements"?
Although we don't do that for MP_MAP (vclock).

File: doc/dev_guide/internals/iproto/images/repl_heartbeat_message.puml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    BackgroundColor transparent
+}
+
+json "**Heartbeat message from master**" as heartbeat_request {
+    "Size": "MP_UINT",
+    "Header": {
+      "[[https://tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general IPROTO_REQUEST_TYPE]]": "IPROTO_OK",
+      "[[https://tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general-replication IPROTO_REPLICA_ID]]": "MP_UINT",
+      "[[https://tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general IPROTO_TIMESTAMP]]": "MP_DOUBLE",
+      "(Optional) [[https://tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#general-replication IPROTO_VCLOCK_SYNC]]": "MP_UINT"
```
Comment (sergepetrenko) [inline doc/dev_guide/internals/iproto/images/repl_heartbeat_message.puml:18]: IPROTO_VCLOCK_SYNC is part of a body, not header.
So the whole "body" is optional in this case and contains a IPROTO_VCLOCK_SYNC

File: doc/dev_guide/internals/iproto/keys.rst [hunk 3]
```diff
@@ -274,10 +286,9 @@ Synchronous replication
             -   Code and |br| value type
             -   Description
 
-        *   -   IPROTO_TERM
+        *   -   :ref:`IPROTO_TERM <internals-iproto-keys-term>`
             -   0x53 |br| MP_UINT
-            -   The term in which the node becomes the :ref:`synchronous queue owner <box_info_synchro>`.
-                Used in :ref:`IPROTO_RAFT_PROMOTE <internals-iproto-replication-raft_promote>`
+            -   The term in which the node becomes the :ref:`synchronous queue owner <box_info_synchro>`
```
Comment (sergepetrenko) [inline doc/dev_guide/internals/iproto/keys.rst]: It's not the term in which the node becomes the synchronous queue owner anymore (because it's used in multiple contexts now). It's just the RAFT term now.

**Review by sergepetrenko — CHANGES_REQUESTED** [review]
Thanks for the patch!

Looks good generally, please find a couple of comments below.

File: doc/dev_guide/internals/iproto/images/repl_ballot.puml (hunk 1 above)
Comment (xuniq) [reply doc/dev_guide/internals/iproto/images/repl_ballot.puml:19]: Updated description

File: doc/dev_guide/internals/iproto/images/repl_heartbeat_message.puml (hunk 2 above)
Comment (xuniq) [reply doc/dev_guide/internals/iproto/images/repl_heartbeat_message.puml:18]: Fixed

File: doc/dev_guide/internals/iproto/keys.rst (hunk 3 above)
Comment (xuniq) [reply doc/dev_guide/internals/iproto/keys.rst]: Updated description

File: doc/dev_guide/internals/iproto/keys.rst [hunk 4]
```diff
@@ -680,6 +713,18 @@ They all have the MP_MAP type:
     It is present only on the instances in the :ref:`"candidate" state <cfg_replication-election_mode>`
     (IPROTO_RAFT_STATE == 2).
 
+..  _internals-iproto-keys-ballot:
+
+IPROTO_BALLOT keys
+~~~~~~~~~~~~~~~~~~
+
+All IPROTO_BALLOT_* keys are only used in :ref:`IPROTO_BALLOT <box_protocol-ballots>` requests.
+There have been some name changes starting with versions :doc:`release/2.7.3`, :doc:`/release/2.8.2`,
```
Comment (andreyaksenov) [inline doc/dev_guide/internals/iproto/keys.rst]: Missing slash here:
```
:doc:`release/2.7.3` -> :doc:`/release/2.7.3`
```

And I'd rephrase `some name changes` to smth else without `some` :) For example, `the following`.

File: doc/dev_guide/internals/iproto/keys.rst [hunk 5]
```diff
@@ -210,28 +216,39 @@ General replication
...
                 Since v. :doc:`2.7.3 </release/2.7.3>` and :doc:`2.8.2 </release/2.8.2>`
+
+        *   -   :ref:`IPROTO_BALLOT_BOOTSTRAP_LEADER_UUID <internals-iproto-keys-ballot>`
+            -   0x08 |br| MP_STR
+            -   UUID of the bootstrap leader. The UUID is encoded as a 36-byte string.
+                Since v. 2.11
+
+        *   -   :ref:`IPROTO_BALLOT_REGISTERED_REPLICA_UUIDS <internals-iproto-keys-ballot>`
+            -   0x09 |br| MP_ARRAY
+            -   An array of MP_STR elements that contains the UUIDs of the registered members in the replica set.
```
Comment (andreyaksenov) [inline doc/dev_guide/internals/iproto/keys.rst]: What do you think about this wording?
> ... of the registered members in the replica set => 
> ... of members registered in the replica set

File: doc/dev_guide/internals/iproto/keys.rst [hunk 6]
```diff
@@ -663,15 +688,23 @@ In fact, it represents the number of logical operations executed on a specific n
...
 
-*   IPROTO_BALLOT_VCLOCK (0x02) is sent in response to :ref:`IPROTO_VOTE <internals-iproto-replication-vote>`.
+*   IPROTO_VCLOCK_SYNC (0x5a) is used by :ref:`replication heartbeats <box_protocol-heartbeat>`.
+    The master sends its heartbeats, including this monotonically growing key, to a replica.
+    Once the replica receives a heartbeat with a non-zero IPROTO_VCLOCK_SYNC value,
+    it starts responding with the same value in all its acknowledgements.
+    This key was introduced in version 2.11.
+
+*   IPROTO_BALLOT_VCLOCK (0x02) is included in the :ref:`IPROTO_BALLOT <box_protocol-ballots>` message.
+    IPROTO_BALLOT is sent in response to :ref:`IPROTO_VOTE <internals-iproto-replication-vote>` request.
```
Comment (andreyaksenov) [inline doc/dev_guide/internals/iproto/keys.rst]: to request -> to the request

File: doc/dev_guide/internals/iproto/keys.rst [hunk 7]
```diff
@@ -663,15 +688,23 @@ In fact, it represents the number of logical operations executed on a specific n
...
     This key was introduced in :doc:`/release/2.6.1`.
 
-*   IPROTO_BALLOT_GC_VCLOCK (0x03) is also sent in response to :ref:`IPROTO_VOTE <internals-iproto-replication-vote>`.
+*   IPROTO_BALLOT_GC_VCLOCK (0x03) is also included in the :ref:`IPROTO_BALLOT <box_protocol-ballots>` message.
+    IPROTO_BALLOT is sent in response to :ref:`IPROTO_VOTE <internals-iproto-replication-vote>` request.
```
Comment (andreyaksenov) [inline doc/dev_guide/internals/iproto/keys.rst]: to request -> to the request

File: doc/dev_guide/internals/iproto/keys.rst [hunk 8]
```diff
@@ -680,6 +713,18 @@ They all have the MP_MAP type:
...
+
+IPROTO_BALLOT keys
+~~~~~~~~~~~~~~~~~~
+
+All IPROTO_BALLOT_* keys are only used in :ref:`IPROTO_BALLOT <box_protocol-ballots>` requests.
+There have been some name changes starting with versions :doc:`release/2.7.3`, :doc:`/release/2.8.2`,
+and :doc:`/release/2.10.0`:
+
+*   IPROTO_BALLOT_IS_RO_CFG (0x01) was formerly called IPROTO_BALLOT_IS_RO.
+*   IPROTO_BALLOT_IS_RO was (0x04) formerly called IPROTO_BALLOT_IS_LOADING.
```
Comment (andreyaksenov) [inline doc/dev_guide/internals/iproto/keys.rst]: Let's use the same structure for both sentences:
> IPROTO_BALLOT_IS_RO_CFG (0x01) was formerly ...
> IPROTO_BALLOT_IS_RO (0x04) was formerly ...

File: doc/dev_guide/internals/iproto/replication.rst [hunk 9]
```diff
@@ -64,12 +64,21 @@ Once in :ref:`replication_timeout <cfg_replication-replication_timeout>` seconds
...
+The master's heartbeat has no body if the IPROTO_VCLOCK_SYNC key is omitted.
+
+The message from master to a replica:
 
 ..  raw:: html
-    :file: images/repl_heartbeat.svg
+    :file: images/repl_heartbeat_message.svg
 
-IPROTO_TIMESTAMP is a float-64 MP_DOUBLE 8-byte timestamp.
+The response from replica:
```
Comment (andreyaksenov) [inline doc/dev_guide/internals/iproto/replication.rst]: from replica -> from the replica

**Review by andreyaksenov — CHANGES_REQUESTED** [review]
Looks good, found one broken link and added several minor suggestions.

File: doc/dev_guide/internals/iproto/keys.rst (hunk 4 above)
Comment (xuniq) [reply doc/dev_guide/internals/iproto/keys.rst]: Fixed

**Review by andreyaksenov — APPROVED** [review]

**Review by sergepetrenko — APPROVED** [review]
Thanks! LGTM.
